pkg/controller/wildflyserver/wildflyserver_controller.go

```diff
@@ -196,63 +306,123 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
 		}
 	}
 
-	// Requeue until the pod list matches the spec's size
-	podList := &corev1.PodList{}
-	labelSelector := labels.SelectorFromSet(labelsForWildFly(wildflyServer))
-	listOps := &client.ListOptions{
-		Namespace:     wildflyServer.Namespace,
-		LabelSelector: labelSelector,
-	}
-	err = r.client.List(context.TODO(), listOps, podList)
-	if err != nil {
-		reqLogger.Error(err, "Failed to list pods.", "WildFlyServer.Namespace", wildflyServer.Namespace, "WildFlyServer.Name", wildflyServer.Name)
-		return reconcile.Result{}, err
-	}
-	size := wildflyServer.Spec.Size
-	if len(podList.Items) != int(size) {
-		reqLogger.Info("Number of pods does not match the desired size", "PodList.Size", len(podList.Items), "Size", size)
-		return reconcile.Result{Requeue: true}, nil
-	}
-
 	// Update WildFly Server host status
-	update := false
 	if r.isOpenShift && !wildflyServer.Spec.DisableHTTPRoute {
 		hosts := make([]string, len(foundRoute.Status.Ingress))
 		for i, ingress := range foundRoute.Status.Ingress {
 			hosts[i] = ingress.Host
 		}
 		if !reflect.DeepEqual(hosts, wildflyServer.Status.Hosts) {
-			update = true
+			updateWildflyServer = true
 			wildflyServer.Status.Hosts = hosts
 			reqLogger.Info("Updating hosts", "WildFlyServer", wildflyServer)
 		}
 	}
 
-	// Update WildFly Server pod status
-	requeue, podsStatus := getPodStatus(podList.Items)
+	if wildflyServer.Status.ScalingdownPods != numberOfPodsToScaleDown {
+		wildflyServer.Status.ScalingdownPods = numberOfPodsToScaleDown
+		updateWildflyServer = true
+	}
+	// Ensure the pod states are up to date by switching it to active when statefulset size follows the wilflyserver spec
+	if numberOfPodsToScaleDown <= 0 {
+		for k, v := range wildflyServer.Status.Pods {
+			if v.State != wildflyv1alpha1.PodStateActive {
+				wildflyServer.Status.Pods[k].State = wildflyv1alpha1.PodStateActive
+				updateWildflyServer = true
+			}
+		}
+	}
+
+	// Update WildFly Server pod status based on the number of StatefulSet pods
+	requeue, podsStatus := getPodStatus(podList.Items, wildflyServer.Status.Pods)
 	if !reflect.DeepEqual(podsStatus, wildflyServer.Status.Pods) {
-		update = true
+		updateWildflyServer = true
 		wildflyServer.Status.Pods = podsStatus
+		reqLogger.Info("Updating the pod status with new status", "Pod statuses", podsStatus)
 	}
 
-	if update {
+	if updateWildflyServer {
 		if err := r.client.Status().Update(context.Background(), wildflyServer); err != nil {
-			reqLogger.Error(err, "Failed to update pods in WildFlyServer status.")
+			reqLogger.Error(err, "Failed to update WildFlyServer status.")
 			return reconcile.Result{}, err
 		}
+		requeue = true
 	}
-	if requeue {
-		return reconcile.Result{Requeue: true}, nil
+
+	mustReconcileStatefulset, requeueStatefulset, err := r.checkStatefulSet(wildflyServer, foundStatefulSet, podList)
```

Comment:
I'm not sure if that is efficient to check update of the StatefulSet after creating all the resources (service, route) that may depend on it.

My idea was to keep the reconcile loop as linear as possible

```
1. check if the statefulset exists
2. if it exists, does it need to be updated?
  2.a yes, update it and requeue
  2.b no, continue
// at this point, we know that the statefulset spec is up to date with the wildflyserver
// any resources will be created from an up to state statefuls
...
```

Reply:
I see, fair point. I will change this.